rust/ql/lib/codeql/rust/internal/TypeInference.qll

```diff
@@ -83,38 +83,44 @@ private module Input1 implements InputSig1<Location> {
 
   int getTypeParameterId(TypeParameter tp) {
     tp =
-      rank[result](TypeParameter tp0, int kind, int id |
+      rank[result](TypeParameter tp0, int kind, int id1, int id2 |
         tp0 instanceof ArrayTypeParameter and
         kind = 0 and
-        id = 0
+        id1 = 0 and
+        id2 = 0
         or
         tp0 instanceof RefTypeParameter and
         kind = 0 and
-        id = 1
+        id1 = 0 and
+        id2 = 1
         or
         tp0 instanceof SliceTypeParameter and
         kind = 0 and
-        id = 2
+        id1 = 0 and
+        id2 = 2
         or
         kind = 1 and
-        id = idOfTypeParameterAstNode(tp0.(DynTraitTypeParameter).getTypeParam())
+        id1 = 0 and
+        id2 = idOfTypeParameterAstNode(tp0.(DynTraitTypeParameter).getTypeParam())
         or
         kind = 2 and
-        exists(AstNode node | id = idOfTypeParameterAstNode(node) |
+        id1 = idOfTypeParameterAstNode(tp0.(ImplTraitTypeParameter).getImplTraitTypeRepr()) and
+        id2 = idOfTypeParameterAstNode(tp0.(ImplTraitTypeParameter).getTypeParam())
```

Comment:
Can you explain what's going on here?  I saw the consistency test failure this fixes, but I'm not really clear about what was wrong.

Reply:
The problem was that an `ImplTraitTypeParameter` is made up of an `ImplTraitTypeRepr` _and_ a `TypeParam`, but only the type parameter was used when creating ids. So in
```rust
fn get_a_my_trait4<T: Clone>(x: T) -> (impl MyTrait<T>, impl MyTrait<T>) {
```
There's two different `ImplTraitTypeParameter`s: one the first `ImplTraitTypeRepr` and one for the second, but they both have the same type parameter (`T`) in thus ended up with the same id. This can cause problems and is what the consistency check complained about.

The fix was to ensure that both AST nodes where considered in the ordering that creates the ids.